[WIP] Add article/zero knowledge architecture (#208)
sogilis/Blog: site/content/posts/2020-11-29-zero-knowledge-architecture.md

---
title: Sécurité et éthique avec l'Architecture Zero Knowledge
author: Mathieu W.
date: 2020-11-20
image: /img/image-name.jpeg
categories:
  - cat1
tags:
  - cysec
  - architecture
  - crypto
  - ZKA
---

## Introduction

> On retrouve une quantité grandissante de données très sensibles stockées dans le cloud. Suite aux différents scandales de fuite d'informations la conscience du grand public sur les problématiques de vie privée en ligne et de sécurité des données n'a cessée de croître. Cela soulève de nombreuses questions, qui stocke mes données ? Qui d'autre peut y avoir accès ? Et pendant le transport de ces données sur le réseau ? 

Reviewer: ```suggestion
On retrouve une quantité grandissante de données très sensibles stockées dans le cloud (documents d'identité, bulletin de salaires en pièce jointe des mails, etc.). Aussi, la possibilité de fuite de ces données n'est plus à prouver aujourd'hui, et la conscience du grand public sur les problématiques de protection de la vie privée et des données sensible n'a cessée de croître. Les questions suivantes gagnent alors en popularité : qui stocke mes données ? Comment ? Qui d'autre peut y avoir accès ? Sont-elles protégées pendant le transport sur le réseau ? 
```

Notes: il ne faut pas citer de scandales si tu ne donne pas des exemples, et je pense que c'est un poil hors propos, alor je suggère de virer.

> Les utilisateurs veulent garder le contrôle sur ses données, savoir qui a accès à quoi et pendant combien de temps.

Reviewer: ```suggestion
En d'autres termes, les utilisateurs veulent garder le contrôle de leur données, savoir qui a accès à quoi et pendant combien de temps.
```

> L'architecture Zero Knowledge (ou ZKA) est un principe d'architecture logicielle qui permet de donner sélectivement l'accès à des données ou de vérifier des conditions sur des données sans avoir à donner l'accès direct aux informations. 

Reviewer: ```suggestion
L'architecture Zero Knowledge (ou ZKA) est un principe d'architecture logicielle visant à assurer la protection et le contrôle des données à caractère privé. Idéalement, les services « zero knowledge » n'ont même pas accès aux données qu'ils manipulent (d'où le principe de connaissance nulle). Par exemple, on peut considérer qu'un service de stockage de données dit « zero knowledge », met en place un système de chiffrement qui assure que seul l'utilisateur final peut lire les dites données.

À noter que le nom « Zero Knowledge Architecture » est issue du concept de « [preuve à divulgation nulle de connaissance](https://fr.wikipedia.org/wiki/Preuve_%C3%A0_divulgation_nulle_de_connaissance) » aka « Zero Knowledge Proof » en anglais, lequel 
```

> C'est une approche architecturale qui met la priorité sur la vie privée des utilisateurs. En intégrant la sécurité profondément dans la conception du logiciel il est possible de de garantir que seul l'utilisateur peut déchiffrer les données, les acteurs de la transmission et du stockage ne peuvent donc pas les exploiter. Le système permet d'utiliser les informations chiffrées sur des services externes cependant sans divulguer plus que le strictement nécessaire.

Reviewer: ```suggestion
Par exemple, on peut considérer qu'un service de stockage de données dit « zero knowledge », met en place un système de chiffrement qui assure que seul l'utilisateur final peut lire les dites données.

À noter que le nom « Zero Knowledge Architecture » est issue du concept de « [preuve à divulgation nulle de connaissance](https://fr.wikipedia.org/wiki/Preuve_%C3%A0_divulgation_nulle_de_connaissance) » aka « Zero Knowledge Proof » en anglais. 

> Cette expression désigne un protocole sécurisé dans lequel une entité, nommée « fournisseur de preuve », prouve mathématiquement à une autre entité, le « vérificateur », qu'une proposition est vraie sans toutefois révéler d'autres informations que la véracité de la proposition
```

## Quelle utilité ? 

> Il y a de nombreux domaines d'applications possibles et certaines entreprises utilisent déjà cette architecture ou certains de ses grands principes dans des produits. Par exemple, **[Signal](https://signal.org/blog/private-contact-discovery/)**, **[NordPass](https://nordpass.com/features/zero-knowledge-architecture/)** ou **[CryptPad](https://blog.cryptpad.fr/2017/02/20/Time-to-Encrypt-the-Cloud/)** qui sont respectivement, une application de messagerie chiffrée, un gestionnaire de mots de passe et un concurrent à google suite. Le lien vers CryptPad est un article contenant une liste d'autre service zero knowledge.

Reviewer: ```suggestion
Il y a de nombreux domaines d'applications possibles pour la ZKA et certaines entreprises utilisent déjà ses principes dans leurs produits. Par exemple, **[Signal](https://signal.org/blog/private-contact-discovery/)**, **[NordPass](https://nordpass.com/features/zero-knowledge-architecture/)** ou **[CryptPad](https://blog.cryptpad.fr/2017/02/20/Time-to-Encrypt-the-Cloud/)** qui sont respectivement, une application de messagerie chiffrée, un gestionnaire de mots de passe et un concurrent à google suite. Le lien vers CryptPad est un article contenant une liste d'autres services zero knowledge.
```

> Etant donné qu'il s'agit uniquement d'un principe d'architecture il pourrait être appliqué à presque tous les domaines, pour mieux comprendre nous allons nous concentrer sur un exemple, jouer au loto en ligne.

Reviewer: ```suggestion
Etant donné qu'il s'agit uniquement d'un principe d'architecture il pourrait être appliqué à presque tous les domaines, pour mieux comprendre nous allons nous concentrer sur un exemple : jouer au loto en ligne.
```

> Pour prouver son âge typiquement on montre une pièce d'identité, après vérification aucune donnée n'est conservée. Cependant sur internet ce n'est pas si simple, et si on en vient à montrer une pièce d'identité, on la numérise puis la transmets. Ceci se fait généralement de manière non chiffrée ce qui représente déjà un problème, d'autant plus que la pièce d'identité contient plus d'informations que nécessaire pour prouver qu'on a le droit de jouer au loto. Mais également après vérification nous ne savons pas vraiment comment sont gérées les données, malgré les RGPD il y a toujours des doutes sur les pratiques de stockage des données. 

Reviewer: ```suggestion
Pour prouver son âge typiquement on montre une pièce d'identité, après vérification aucune donnée n'est conservée. Cependant sur Internet ce n'est pas si simple, et si on en vient à montrer une pièce d'identité, on la numérise puis la transmet. Ceci se fait généralement de manière non chiffrée, ce qui représente un premier problème, laissant les données d'identité de l'utilisateur, sensibles à des écoutes ou a des fuites sur les canaux de communucation. Un second problème s'ajoute à celà : la pièce d'identité contient plus d'informations que nécessaire. En effet, il suffit de prouver son âge afin d'obtenir le droit de jouer au loto, mais dans le cas de la pièce d'identité, ce sont bien toutes les données d'identités (prénom, nom, âge, mais aussi adresse postale et nationalité) qui sont révélées. Ce second problème est aggravé par le fait que nous ne savons pas vraiment comment sont gérées les données de la pièce d'indentité, une fois la preuve de majorité obtenue. Malgré l'application de la loi RGPD, des doutes sur les pratiques de stockage des données subsistent, en l'absence d'audit complet de l'infrastructure du service.
```

> Maintenant imaginons qu'il existe une application comme le **[Dossier Médical Partagé](https://www.dmp.fr/)** qui gère le suivi médical et l'identité en adhérant aux principes zero knowledge. Si Billy veut s'inscrire sur un site de loto en ligne il doit prouver qu'il est majeur, dans le cas où le site accepte ce mode d'identification Billy peut le faire sans dévoiler son âge exact ou son identité. Etant donné que le système contient aussi ses informations médicales le site du loto pourrait utiliser l'API pour vérifier que Billy n'a pas d'historique d'addiction aux jeux d'argent avant de lui permettre de jouer de grosses sommes. L'accès aux données sera nécessairement validé par Billy, et son historique médical exact ne sera pas révélé, l'application répondra avec un score abstrait ou un booléen. 

Reviewer: ```suggestion
Imaginons maintenant un second scénario, tel qu'une application médicale ayant accès au **[Dossier Médical Partagé](https://www.dmp.fr/)**. Cette application gère le suivi médical et l'identité des patients, en adhérant aux principes zero knowledge.
```

la suite de ce paragraphe manue cruellement de détail et d'explications, c'et bourréde sous-entendus (et il en faut aucun quand on rédigie un article): 

`Si Billy veut s'inscrire sur un site de loto en ligne il doit prouver qu'il est majeur, dans le cas où le site accepte ce mode d'identification Billy peut le faire sans dévoiler son âge exact ou son identité. Etant donné que le système contient aussi ses informations médicales le site du loto pourrait utiliser l'API pour vérifier que Billy n'a pas d'historique d'addiction aux jeux d'argent avant de lui permettre de jouer de grosses sommes. L'accès aux données sera nécessairement validé par Billy, et son historique médical exact ne sera pas révélé, l'application répondra avec un score abstrait ou un booléen. `

## Axes de mise en oeuvre

  1. Authentification à preuve à divulgation nulle de connaissance
  2. Approche non naïve
  3. Chiffrement de bout en bout (ou E2EE pour End to End Encryption)
  
**Authentification à preuve à divulgation nulle de connaissance**

> S'authentifier avec une comabinaison de nom d'utilisateur et de mot de passe a plusieurs inconvénients, dans ce système l'authentification est faite sans transférer de mot de passe, ce dernier protège localement un certificat et une paire de clés, une de chiffrement et une de signature dédiée à prouver son identité auprès du serveur.

Reviewer: J'attends le schéma, mais j'aimerai bien un peu plus d'explications, maintenant qu'on a dit que ceprincipe était au coeur de la ZKA. 

Après, l'introd de l'article Wikipedia est pas mal je trouve (je l'ai cité plus haut): 
> Une preuve à divulgation nulle de connaissance est une brique de base utilisée en cryptologie dans le cadre de l'authentification et de l'identification. Cette expression désigne un protocole sécurisé dans lequel une entité, nommée « fournisseur de preuve », prouve mathématiquement à une autre entité, le « vérificateur », qu'une proposition est vraie sans toutefois révéler d'autres informations que la véracité de la proposition.

 >En pratique, ces schémas se présentent souvent sous la forme de protocoles de type « défi/réponse » (challenge-response)1. Le vérificateur et le fournisseur de preuve s'échangent des informations et le vérificateur contrôle si la réponse finale est positive ou négative.

SCHEMA FONCTIONNEL 

**Approche non naÏve**

> Typiquement lorsqu'on stock des données réservées à un utilisateur on utilise un paradigme de base de données NoSQL, ce qui nous donne un document par utilisateur. Compte tenu de nos objectifs en terme de vie privée on ne va pas tout transmettre quand quelqu'un nous demande l'accès à certaines données, c'est une approche dite "non naïve". On utilie des blobs, ce sont des ensembles de données liées et les plus unitaires possibles. Des exemples de blobs peuvent être le blob "identité", qui contient uniquement votre identité, le blob "est majeur(e) ?" pour donner l'accès à cette information uniquement. Ou encore le blob "historique de rendez vous médicaux" qui lui même contient un blob "informations de facturation". Cela permet de contrôler avec une granularité extrêmement fine les données que l'on va partager avec les différents services qui ont des besoins différents. 

Reviewer: ```suggestion
Compte tenu de nos objectifs en terme de vie privée on ne va pas tout transmettre quand quelqu'un nous demande l'accès à certaines données, c'est une approche dite "non naïve". On utilie des blobs, ce sont des ensembles de données liées et les plus unitaires possibles. Des exemples de blobs peuvent être le blob "identité", qui contient uniquement votre identité, le blob "est majeur(e) ?" pour donner l'accès à cette information uniquement. Ou encore le blob "historique de rendez vous médicaux" qui lui même contient un blob "informations de facturation". Cela permet de contrôler avec une granularité extrêmement fine les données que l'on va partager avec les différents services qui ont des besoins différents. 
```

Je ne comprend pas l'intérêt de la première phrase, ni ce quele paradigme NoSQL vient faire là...

**Chiffrement de bout en bout**

> Toutes les opérations de chiffrement et déchiffrement sont réalisées côté client, c'est à dire sur la machine de l'utilisateur. C'est également le cas pour partager des données avec un service externe, les données ne sont jamais stockées ou transmises en clair. Lorsqu'un service requête des données c'est seulement une fois arrivées au service qui possède la clé privée que les données pourront être lues. Non seulement chaque service qui consomme vos données a une clé différente mais il est possible pour chaque service de contrôler l'accès à chaque blob individuellement. Pour les performances il est préférable de faire de l'encapsulation de clé symétrique. Le serveur central qui sert de tiers de confiance peut révoquer le certificat d'un service pour bloquer l'accès, l'utilisateur a également plein controle sur l'accès à ses données.

Reviewer: ```suggestion
Idéalement, dans une architecture zero-knowledge, outes les opérations de chiffrement et déchiffrement sont réalisées côté client, c'est à dire sur la machine de l'utilisateur. C'est également le cas pour partager des données avec un service externe, les données ne sont jamais stockées ou transmises en clair. Lorsqu'un service requête des données c'est seulement une fois arrivées au service qui possède la clé privée que les données pourront être lues. Non seulement chaque service qui consomme vos données a une clé différente mais il est possible pour chaque service de contrôler l'accès à chaque blob individuellement. Pour les performances il est préférable de faire de l'encapsulation de clé symétrique. Le serveur central qui sert de tiers de confiance peut révoquer le certificat d'un service pour bloquer l'accès, l'utilisateur a également plein controle sur l'accès à ses données.
```

> ## Points forts

Reviewer: ```suggestion
## Points forts de La ZKA
```

> Il est possible de développer des clients pour ce type d'architecture sur toutes les plateformes, en clients lourds type applications mobiles ou application desktop, mais également dans le navigateur en utilisant les APIs récents.

Reviewer: ```suggestion
L'écosystème de bibliothèque de cryptographie permet de développer tout types de clients : client lourds, mobiles ou Web, pour les services zero-knowledge.
```

> Cette architecture présente très peu de risques pour les données stockées, si les données fuitent du serveur elles sont chiffrées et donc sans valeur. Pour la même raison si quelqu'un a des privilèges sur la machine qui contient les données il n'a pas accès à nos données en clair puisque l'utilisateur est le seul à posséder la clé nécessaire au déchiffrement. 

Reviewer: ```suggestion
Cette architecture présente très peu de risques pour les données stockées, si les données fuitent du serveur elles sont chiffrées et donc sans valeur. Pour la même raison si quelqu'un a des privilèges sur la machine qui contient les données il n'a pas accès à nos données en clair puisque l'utilisateur est le seul à posséder la clé nécessaire au déchiffrement. 
```

> Cette architecture présente très peu de risques pour les données stockées, si les données fuitent du serveur elles sont chiffrées et donc sans valeur.

:warning: C'est faux ! Une fuite de données critiques, même fortement chiffrées, n'est pas considérée comme « sans valeur »

> L'authentification zero knowledge permet de prévenir un bon nombre d'attaque qui viserait à espionner et usurper les identifiants de l'utilisateur. Le mot de passe ne transite jamais sur le réseau et le challenge émis par le serveur est différent à chaque fois.

Reviewer: ```suggestion
L'authentification zero knowledge permet de prévenir un bon nombre d'attaque qui viserait à espionner et usurper les identifiants de l'utilisateur. En effet, la systématisation du principe de défi cryptographique, implique qu'il n'y a plus d'échange de mot de passe sur le réseau, ce qui diminue le risque de fuite de données d'authentification. De plus, le challenge émis par le serveur étant différent à chaque requête, les attaques de types « man in the middle » sont aussi rendues plus difficiles à opérer.
```

> Le chiffrement de bout en bout résout le manque de confiance en le serveur et les acteurs de transmission qui ne peuvent pas accéder aux données en clair.

Reviewer: Pas compris :thinking: Tu veux dire qu'on peut faire confiance au réseau ? 

Du coup je propose de prendre ma suggestion sur les attaques type « man in the middle » et virer ce paragraphe.

> Préservation de la vie privée en ne divulguant que les données nécessaire et ce avec l'accord de l'utilisateur, qui est révocable à tout moment.

Reviewer: ```suggestion
Préservation de la vie privée en ne divulguant que les données nécessaires, et ce, avec l'accord de l'utilisateur, qui est révocable à tout moment.
```

## Limites 

> Compte tenu de l'omniprésence de la cryptographie dans l'architecture elle est assez complexe ce qui la rend coûteuse à mettre en place. C'est particulièrement le cas à cause du manque (temporaire ) de framework ou autre librairies pour faciliter le développement.

Reviewer: ```suggestion
Compte tenu de l'omniprésence de la cryptographie dans l'architecture, la ZKA est assez complexe, ce qui implique une montée en compétence sur la cryptographie et un coût de mise en place supplèmentaire.
```

> Dans les cas où il faut absolument être sur de pouvoir retrouver ses données il faut mettre en place un système de récupération de clé, mais ce n'est pas critique tout le temps.

Reviewer: ```suggestion
Dans les cas où il faut absolument être sur de pouvoir retrouver ses données il faut mettre en place un système de récupération de clé, mais ce n'est pas critique tout le temps.
```

Pas compris :thinking:  

à raffiner ensemble.

> Si on veut développer le client dans les navigateurs on a besoin de nombreuses mesures de sécurités, CORS, CSP (Content security policy), SRI (verification d'assets), Referrer-Policy et File-API, et idéalement faire tourner la couche crypto en WebAssembly afin de compliquer les manipulations lors de l'exécution.

Reviewer: Un trollsu Javascript est caché dans ce paragrapghe (je prends :smile: )

Mais il faudra reformuler aussi

> Il faut avoir confiance en l'implémentation de l'application ZKA, il faut au minimum que le code soit audité et idéalement que le code soit open source afin de garantir l'absence d'erreurs ou de backdoors.

Reviewer: ```suggestion
Afin d'avoir confiance en l'implémentation de l'application ZKA, il faut au minimum que le code soit audité par un tiers de confiance, ou ouvert (licence libre ou open-source). De cette façon, on a un moyen d'étudier l'absence d'erreurs ou de backdoors.
```

> On atteint jamais le zero knowledge, le fonctionnement de l'application nécessite forcément la connaissance d'une adresse IP et la clé publique de chiffrement qui peut être l'identifiant, pas moyen d'échapper à ça. Or cela engendre des métadonnées qui peuvent déjà laisser transparaître des informations.   

Reviewer: Comme expliqué dans l'article ...,

 ## Conclusion

> L'évolution des lois et de la mentalité des utilisateurs ammène des principes d'architecture logicielle radicalement nouveaux, centrés sur la vie privée et la sécurité. L'apparition de nouveaux services basés sur cette architecture va stimuler l'écosystème ce qui développera à terme les briques technologiques nécessaires pour que toute l'industrie puisse appliquer ces approches architecturales. Les utilisateurs ont beaucoup à y gagner puisqu'en plus de la sécurité accrue, ces pratiques assurent l'éthique des développeurs qui ne peuvent pas accéder aux informations afin de faire de la publicité ciblée ou d'espionner leurs utilisateurs.

Reviewer: On peut retravailler un peu la conclusion aussi je pense :)

---
title: Sécurité et éthique avec l'Architecture Zero Knowledge

> author: Mathieu W.

Reviewer: @m-wy ça te dérange si on ajout Jordan Prudhomme et Willy Malvault ?

Author: No problem

date: 2020-11-20

> image: /img/image-name.jpeg

Reviewer: Il faut trouver une image :)

On peut chercher là pour des images avec licence CC-by https://search.creativecommons.org/photos/9ed7de27-8c9b-4aa1-9bae-62883d23302f

categories:
  - cat1
tags:
  - cysec
  - architecture
  - crypto
  - ZKA
  - data privacy
  - cybersecurity
---

## Introduction

> On retrouve une quantité grandissante de données très sensibles stockées dans le cloud (documents d'identité, bulletin de salaires en pièce jointe des mails, etc.). Aussi, la possibilité de fuite de ces données n'est plus à prouver aujourd'hui, et la conscience du grand public sur les problématiques de protection de la vie privée et des données sensibles n'a cessée de croître. Les questions suivantes gagnent alors en popularité : qui stocke mes données ? Comment ? Qui d'autre peut y avoir accès ? Sont-elles protégées pendant le transport sur le réseau ? 

Reviewer: ```suggestion
Aujourd'hui, on retrouve une quantité grandissante de données sensibles, privées et personelles stockées dans le cloud (documents d'identité, bulletins de salaires dans les emails, etc.). Aussi, la possibilité de fuite de ces données n'est plus à prouver aujourd'hui, et la conscience du grand public sur les problématiques de protection de la vie privée et des données sensibles n'a cessée de croître. Les questions suivantes gagnent alors en popularité : qui stocke mes données ? Comment ? Qui d'autre peut y avoir accès ? Sont-elles protégées pendant le transport sur le réseau ? 
```

> En d'autres termes, les utilisateurs veulent garder le contrôle de leur données, savoir qui a accès à quoi et pendant combien de temps.

Reviewer: ```suggestion
En d'autres termes, les utilisateurs veulent garder le contrôle de leur données, savoir qui a accès, à quoi, et pendant combien de temps.
```

L'architecture Zero Knowledge (ou ZKA) est un principe d'architecture logicielle visant à assurer la protection et le contrôle des données à caractère privé. Idéalement, les services « zero knowledge » n'ont même pas accès aux données qu'ils manipulent (d'où le principe de connaissance nulle).
Par exemple, on peut considérer qu'un service de stockage de données dit « zero knowledge », met en place un système de chiffrement qui assure que seul l'utilisateur final peut lire les dites données.

À noter que le nom « Zero Knowledge Architecture » est issu du concept de « [preuve à divulgation nulle de connaissance](https://fr.wikipedia.org/wiki/Preuve_%C3%A0_divulgation_nulle_de_connaissance) » aka « Zero Knowledge Proof » en anglais.

> Cette expression désigne un protocole sécurisé qui est une brique de base utilisée en cryptologie pour prouver qu'une proposition est vraie sans révéler d'autre information que la véracité de la proposition.

Reviewer: ```suggestion
Cette expression désigne un protocole sécurisé, qui est une brique de base utilisée en cryptologie, pour prouver qu'une proposition est vraie, le tout sans révéler d'autre information que la véracité de la proposition.
```

Cette définition ancre le principe suivant : les services d'une architecture dite « Zero knowledge », y compris les éventuels moyen de transport de données qui la composent, ne peuvent accèder à des données utilisateurs, que si le consentement a été explicitement donné.  Nous verrons aussi un peu plus tard, que la granularité des blocs d'informations qui composent les données à caractère privé, joue un rôle essentiel dans les architectures zero knowledge.

## Quelle utilité ? 

> Il y a de nombreux domaines d'applications possibles pour la ZKA et certaines entreprises utilisent déjà ses principes dans leurs produits. Par exemple, **[Signal](https://signal.org/blog/private-contact-discovery/)**, **[NordPass](https://nordpass.com/features/zero-knowledge-architecture/)** ou **[CryptPad](https://blog.cryptpad.fr/2017/02/20/Time-to-Encrypt-the-Cloud/)** qui sont respectivement, une application de messagerie chiffrée, un gestionnaire de mots de passe et un concurrent à google suite. Le lien vers CryptPad est un article contenant une liste d'autres services zero knowledge.

Reviewer: ```suggestion
Il y a de nombreux domaines d'applications possibles pour la ZKA et certaines entreprises utilisent déjà ses principes dans leurs produits. Par exemple, **[Signal](https://signal.org/blog/private-contact-discovery/)**, **[NordPass](https://nordpass.com/features/zero-knowledge-architecture/)** ou **[CryptPad](https://blog.cryptpad.fr/2017/02/20/Time-to-Encrypt-the-Cloud/)** qui sont respectivement, une application de messagerie chiffrée, un gestionnaire de mots de passe et un concurrent à google suite (le lien vers l'article CryptPad fournit une liste d'autres services zero knowledge).
```

> Etant donné qu'il s'agit uniquement d'un principe d'architecture il pourrait être appliqué à presque tous les domaines, pour mieux comprendre nous allons nous concentrer sur un exemple : jouer au loto en ligne.

Reviewer: ```suggestion

Etant donné qu'il s'agit uniquement d'un principe d'architecture il pourrait être appliqué à presque tous les domaines, pour mieux comprendre nous allons nous concentrer sur un exemple : jouer au loto en ligne.
```

> Pour prouver son âge typiquement on montre une pièce d'identité, après vérification aucune donnée n'est conservée. Cependant sur Internet ce n'est pas si simple, et si on en vient à montrer une pièce d'identité, on la numérise puis la transmet. Ceci se fait généralement de manière non chiffrée, ce qui représente un premier problème, laissant les données d'identité de l'utilisateur, sensibles à des écoutes ou a des fuites sur les canaux de communucation. Un second problème s'ajoute à celà : la pièce d'identité contient plus d'informations que nécessaire. En effet, il suffit de prouver son âge afin d'obtenir le droit de jouer au loto, mais dans le cas de la pièce d'identité, ce sont bien toutes les données d'identités (prénom, nom, âge, mais aussi adresse postale et nationalité) qui sont révélées. Ce second problème est aggravé par le fait que nous ne savons pas vraiment comment sont gérées les données de la pièce d'indentité, une fois la preuve de majorité obtenue. Malgré l'application de la loi RGPD, des doutes sur les pratiques de stockage des données subsistent, en l'absence d'audit complet de l'infrastructure du service.

Reviewer: ```suggestion
### Jouer au loto en ligne

Pour prouver son âge, typiquement, on montre une pièce d'identité, après vérification aucune donnée n'est conservée. Cependant sur Internet ce n'est pas si simple, et si on en vient à montrer une pièce d'identité, on la numérise puis la transmet. Ceci se fait généralement de manière non chiffrée, ce qui représente un premier problème, laissant les données d'identité de l'utilisateur, sensibles à des écoutes ou a des fuites sur les canaux de communucation. Un second problème s'ajoute à celà : la pièce d'identité contient plus d'informations que nécessaire. En effet, il suffit de prouver son âge afin d'obtenir le droit de jouer au loto, mais dans le cas de la pièce d'identité, ce sont bien toutes les données d'identités (prénom, nom, âge, mais aussi adresse postale et nationalité) qui sont révélées. Ce second problème est aggravé par le fait que nous ne savons pas vraiment comment sont gérées les données de la pièce d'indentité, une fois la preuve de majorité obtenue. Malgré l'application de la loi RGPD, des doutes sur les pratiques de stockage des données subsistent, en l'absence d'audit complet de l'infrastructure du service.
```

Reviewer: Il manque un paragraphe: 

« Avec une architecture zero-knowledge, un utilisateur pourrait utiliser un service tiers de preuve de majorité, et ne laisserait aucune données sensible transiter sur le réseau. » , ensuite il faudrait décrire un peut comment ça fonctionnerait, peut-être avec un schéma aussi :)

> Pour répondre à ce scénario imaginons une application médicale comme le **[Dossier Médical Partagé](https://www.dmp.fr/)**. Cette application gère le suivi médical et l'identité des patients, en adhérant aux principes zero knowledge. Billy a un compte sur lequel se trouve tout son historique médical numérisé, ses rendez-vous, ses ordonnances etc... Grâce à ce compte il partage sélectivement ses informations avec les professionnels de santé, les ordonnances avec les pharmaciens, les informations de facturation avec sa mutuelle et chaque professionel ne voit que les données qui lui sont utiles. Sur le même principe, si le site du Loto peut accdeder à l'âge et l'identité de Billy il peut s'authentifier et jouer au Loto sans devoir partager de pièce d'identité.

Reviewer: ```suggestion
Intéressons nous à un second scénario bien populaire : le dossier médical partagé.

### Dossier médical partagé
Pour répondre à ce scénario imaginons une application médicale comme le **[Dossier Médical Partagé](https://www.dmp.fr/)**. Cette application gère le suivi médical et l'identité des patients, en adhérant aux principes zero knowledge. Billy a un compte sur lequel se trouve tout son historique médical numérisé, ses rendez-vous, ses ordonnances etc... Grâce à ce compte il partage sélectivement ses informations avec les professionnels de santé, les ordonnances avec les pharmaciens, les informations de facturation avec sa mutuelle et chaque professionel ne voit que les données qui lui sont utiles. Sur le même principe, si le site du Loto peut accdeder à l'âge et l'identité de Billy il peut s'authentifier et jouer au Loto sans devoir partager de pièce d'identité.
```

Reviewer: tu mélanges les deux scénarios :exploding_head:

Il est également possible de ne pas partager l'information directement, on demande alors à un tiers de confiance de valider une condition pour nous, ce qui permet de verifier que quelqu'un est majeur sans apprendre son age exact par exemple.

> ![Schéma fonctionnel d'un example de ZKA](/img/ZKA/zka-loto.gif)

Reviewer: Si tu fais un Gif il faut que tu ajoutes sur les slides un indicateurs de positionnement, par exemple : `1/5` pour le permier slide parmis `5`, puis `2/5`, ...etc ou alors carrèment `étape 1, étapes 2, étapes 3` ...

## Axes de mise en oeuvre

> Cette architecture repose sur 3 grand axes de mises en oeuvre

Reviewer: ```suggestion
La ZKA repose sur 3 axes
```

>   1. Authentification à preuve à divulgation nulle de connaissance

Reviewer: ```suggestion
  1. Utilisation de preuve à divulgation nulle de connaissance
```

Reviewer: On ne l'utilise pas que pour de l'authentification, d'ailleurs dans ton scénario de loto ça n'est pas le cas et c'est bien une prevue à divulgation nulle de connaissance.

  2. Approche non naïve
  3. Chiffrement de bout en bout (ou E2EE pour End to End Encryption)
  

> **Authentification à preuve à divulgation nulle de connaissance**

Reviewer: ```suggestion
**Preuve à divulgation nulle de connaissance**
```

> S'authentifier avec une comabinaison de nom d'utilisateur et de mot de passe a plusieurs inconvénients, dans ce système l'authentification est faite sans transférer de mot de passe.

Reviewer: Paragraphe à déplacer après l'explication je pense.

Pour cela on utilise une **[preuve à divulgation nulle de connaissance](https://fr.wikipedia.org/wiki/Preuve_%C3%A0_divulgation_nulle_de_connaissance)**, c'est un protocole sécurisé dans lequel une entité, nommée « fournisseur de preuve », prouve mathématiquement à une autre entité, le « vérificateur », qu'une proposition est vraie sans révéler d'autres informations que la véracité de la proposition. Généralement ce type de preuve repose sur des protocoles défi/réponse, le "vérficateur" envoie un challenge au "fournisseur de preuve" qui répond grâce à une information que seul lui connait afin de prouver son identité.

**Approche non naÏve**

> Compte tenu de nos objectifs en terme de vie privée on ne va pas tout transmettre quand quelqu'un nous demande l'accès à certaines données, c'est une approche dite "non naïve". On utilie des blobs, ce sont des ensembles de données liées et concùs pour être les plus unitaires possibles. Des exemples de blobs peuvent être le blob "identité", qui contient uniquement votre identité, le blob "est majeur(e) ?" pour donner l'accès à cette information uniquement. Ou encore le blob "rendez-vous médical" qui lui même contient, entres autres, un sous-blob "informations de facturation". Cela permet de contrôler avec une granularité extrêmement fine les données que l'on va partager avec les différents services qui ont des besoins différents. 

Reviewer: ```suggestion
Si un utilisateur a des objectifs de protection de vie privée, il ne doit transmettre que les informations strictement nécessaires, quand un tiers lui demande l'accès à certaines données, c'est une approche dite "non naïve". Pour ce faire, la ZKA utilise des blobs, qui sont des ensembles de données liées et conçus pour être les plus unitaires possibles. Des exemples de blobs peuvent être le blob "identité", qui contient uniquement votre identité, le blob "est majeur(e) ?" pour donner l'accès à cette information uniquement. Ou encore le blob "rendez-vous médical" qui lui même contient, entres autres, un sous-blob "informations de facturation". Cela permet de contrôler avec une granularité extrêmement fine les données que l'on va partager avec les différents services qui ont des besoins différents. 
```

Reviewer: Il faudrait peut-être donner un exemple de ce que serait une approche naïve ?

**Chiffrement de bout en bout**

> Idéalement, dans une architecture zero-knowledge, toutes les opérations de chiffrement et déchiffrement sont réalisées côté client, c'est à dire sur la machine de l'utilisateur. C'est également le cas pour partager des données avec un service externe, les données ne sont jamais stockées ou transmises en clair. Lorsqu'un service demande des données à caractère privé, seul le service identifié comme destinataire possède la clef pour déchiffrer les données transmises.

Reviewer: ```suggestion
Dans une architecture zero-knowledge, tout transfert de données est chiffré. Idéalement, tout stockage de données, même temporaire, est également chiffré. Lorsqu'un service demande des données à caractère privé, seul le service identifié comme destinataire possède la clef pour déchiffrer les données transmises.
```

> Dans le cas ou la source de donnée utilisée n'est pas chiffrée on met en place une pratique proche, le **[Zero Access Encryption](https://protonmail.com/blog/zero-access-encryption/)**.

Reviewer: ```suggestion
**zero-access-encryption**
TODO: décrire
Dans le cas ou la source de donnée utilisée n'est pas chiffrée on met en place une pratique proche, le **[Zero Access Encryption](https://protonmail.com/blog/zero-access-encryption/)**.
```

Je propose de laisser @RaboliotLeGris nous rédiger ce paragraphe, c'est inutile de citer le principe dans au moins le résumer je pense.

## Points forts de La ZKA

> L'écosystème de bibliothèque de cryptographie permet de développer tout types de clients : client lourds, mobiles ou Web, pour les services zero-knowledge.

Reviewer: J'inverserai les 2 premiers paragraphes, dans le sens où on s'en fout un peu de l'écosustème, si on ne sait pas pourquoi c'est bien d'utiliser de la crypto. Non ?

> Cette architecture généralise l'usage du chiffrement des données et de la cryptographie, ce qui apporte un  bénéfice double : premièrement un renforcement de la sécurité des données à caractères privé et deuxiemement cela éduque les utilisateurs à l'usage de la cryptographie.

Reviewer: ```suggestion
Cette architecture généralise l'usage du chiffrement des données et de la cryptographie, ce qui apporte un  bénéfice double :  un renforcement de la sécurité des données à caractères privé, et l'éducation des utilisateurs à l'usage de la cryptographie.
```

> L'authentification zero knowledge permet de prévenir un bon nombre d'attaque qui viserait à espionner et usurper les identifiants de l'utilisateur. En effet, la systématisation du principe de défi cryptographique, implique qu'il n'y a plus d'échange de mot de passe sur le réseau, ce qui diminue le risque de fuite de données d'authentification. De plus, le challenge émis par le serveur étant différent à chaque requête, les attaques de types « man in the middle » sont aussi rendues plus difficiles à opérer.

Reviewer: ```suggestion
La systématisation du principe de défi cryptographique, implique que les échanges de mot de passe sont limités au fournisseur de preuve, ce qui tend à diminuer le risque de fuite de mot de passe. 
De plus, le challenge émis par le serveur étant différent à chaque requête, les tentatives d'usurpation de types « man in the middle » sont aussi rendues plus difficiles à opérer.
```

> Préservation de la vie privée en ne divulguant que les données nécessaires, et ce, avec l'accord de l'utilisateur, qui est révocable à tout moment.

Reviewer: ```suggestion
Préservation de la vie privée avec l'approche non-naïve, en ne divulguant que les données nécessaires. Certains service Zero-knowledge proposent même un contrôle dans le temps d'accès aux données de l'utiliateur, en mettant en place des mécanismes de révocation.
```

## Limites 

> Compte tenu de l'omniprésence de la cryptographie dans l'architecture, la ZKA est assez complexe, ce qui implique une montée en compétence sur la cryptographie et un coût de mise en place supplèmentaire.

Reviewer: J'inverserai les 2 premiers paragraphe, dans le sens où on s'en fout un peu de l'écosustème, si on ne sait pas pourquoi c'est bien d'utiliser de la crypto. Non ?

> Il n'y a pas moyen de réinitialier le mot de passe ou les clés, pour s'assurer de conserver l'accès il faut mettre en place un système de récupération de clé. Ce n'est pas forcément critique, un système de messagerie instantanée par exemple peut se permettre de perde l'historique de conversation cepedant il est inconcevable de perdre l'accès à ses données médicales. 

Reviewer: ```suggestion
Les mécanismes de cryptographies utiliseés par la ZKA sont pus complexe à opérer que les mécanismes classiques. En effet, la cryptographie moderne implique de savoir gérer une clef privée. Par exemple, pour s'assurer de conserver l'accès à des données, il faut mettre en place un système de récupération de clé, là où classiquement on utiliserait une ré-initialisation de mot de passe.

 Attention toutefois à ne pas généraliser cette difficulté. En effet, la récupération de données n'est pas forcément critique pour tous les scénarios, par exemple un service de messagerie instantanée  pourra se permettre de perde l'historique des conversations (et on pourra négliger la récupération des clefs privées utilisées pour le chiffrement des messages).
 À contrario, il est inconcevable de perdre l'accès à ses données médicales. 
```

Pour developper un client ZKA dans les navigateurs on a besoin de nombreuses mesures de sécurités, CORS, CSP (Content security policy), SRI (verification d'assets), Referrer-Policy et File-API. Afin de complexifier les manipulations lors de l'exécution il est preferable de faire tourner la couche crypto en WebAssembly car le javascript peut être manipulé à la volée .

Afin d'avoir confiance en l'implémentation de l'application ZKA, il faut au minimum que le code soit audité par un tiers de confiance, ou ouvert (licence libre ou open-source). De cette façon, on a un moyen d'étudier l'absence d'erreurs ou de backdoors.

> On atteint jamais le zero knowledge, le fonctionnement de l'application nécessite forcément la connaissance d'une adresse IP et la clé publique de chiffrement qui peut être l'identifiant, pas moyen d'échapper à ça comme nous l'explique **[cet article](https://blog.cryptpad.fr/2017/07/07/cryptpad-analytics-what-we-cant-know-what-we-must-know-what-we-want-to-know/)**. Les métadonnées générées par le fonctionnement peuvent déjà laisser transparaître des informations.   

Reviewer: ```suggestion
Dans la vraie vie, on atteint rarement le zero knowledge pur. En effet, le fonctionnement des services nécessite bien souvent la connaissance d'une adresse IP et d'une clé publique de chiffrement, pouvant servir d'identifiant. Il est très compliqué d'échapper à se partage d'informations, comme nous l'explique **[cet article de cryptpad](https://blog.cryptpad.fr/2017/07/07/cryptpad-analytics-what-we-cant-know-what-we-must-know-what-we-want-to-know/)**. Les métadonnées générées par le fonctionnement peuvent déjà laisser transparaître des informations.   
```

Dans la ZKA, on préconise l'utilisation de blobs, qui sont des ensembles de données conçus pour être les plus unitaires possibles. Des exemples de blobs peuvent être le blob "identité", qui contient uniquement l'identité de l'utilisateur, ou encore le blob "est majeur(e) ?" pour donner l'accès à cette information uniquement. Et si on fait la supposition que l'on dispose d'un tiers de confiance, reconnu comme autorité nationale de gestion des données d'identité, lequel disposerait d'un certificat TLS (à tout hasard), on pourrait même signer ce blob « est majeur » afin de justifier de son authenticité. Ce blob signé devient alors une **preuve de majorité à divulgation nulle de connaissance**.

### Preuve à divulgation nulle de connaissance

> La **[preuve à divulgation nulle de connaissance](https://fr.wikipedia.org/wiki/Preuve_%C3%A0_divulgation_nulle_de_connaissance)**, pilier de la ZKA, est un protocole sécurisé dans lequel une entité, nommée « fournisseur de preuve », prouve mathématiquement à une autre entité, le « vérificateur », qu'une proposition est vraie sans révéler d'autres informations que la véracité de la proposition. Généralement ce type de preuve repose sur des protocoles défi/réponse, le "vérficateur" envoie un challenge au "fournisseur de preuve" qui répond grâce à une information que seul lui connait afin de prouver son identité.

Reviewer 2: ```suggestion
La **[preuve à divulgation nulle de connaissance](https://fr.wikipedia.org/wiki/Preuve_%C3%A0_divulgation_nulle_de_connaissance)**, pilier de la ZKA, est un protocole sécurisé dans lequel une entité, nommée « fournisseur de preuve », prouve mathématiquement à une autre entité, le « vérificateur », qu'une proposition est vraie sans révéler d'autres informations que la véracité de la proposition. Généralement ce type de preuve repose sur des protocoles défi/réponse, le "vérificateur" envoie un challenge au "fournisseur de preuve" qui répond grâce à une information que seul lui connaît afin de prouver son identité.
```

Avec une architecture zero-knowledge, l'utilisateur pourrait utiliser un service tiers de preuve de majorité, et ne laisserait aucune données sensible transiter sur le réseau. 
Le fonctionnement du système est schématisé dans l'image ci dessous.

![Schéma fonctionnel d'un example de ZKA](/img/ZKA/zka-loto.gif)

## Authentification zero-knowledge avec le protocole SRP

> S'authentifier avec une comabinaison de nom d'utilisateur et de mot de passe a plusieurs inconvénients, dans ce système l'authentification est faite sans transférer le mot de passe.

Reviewer 2: ```suggestion
S'authentifier avec une combinaison de nom d'utilisateur et de mot de passe a plusieurs inconvénients, dans ce système l'authentification est faite sans transférer le mot de passe.
```

Un exemple d'implémentation de preuve à divulgation nulle est le **[Secure Remote Password protocol (SRP)](https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol)** qui permet de réaliser une authentification sans que le mot de passe ne sorte du client. La **[RFC 5054](https://tools.ietf.org/html/rfc5054)** et la **[RFC 2945](https://www.ietf.org/rfc/rfc2945.txt)** décrivent le fonctionnement de ce protocole en détail.

### La messagerie zero-knowledge Protonmail

**Zero access encryption**
TODO: décrire
Dans le cas ou la source de donnée utilisée n'est pas chiffrée on met en place une pratique proche, le **[Zero Access Encryption](https://protonmail.com/blog/zero-access-encryption/)**.

### Réflexion sur le dossier médical partagé

> C'est un cas d'application idéal pour l'architecture ZKA, d'autant plus l'applicatif actuel du **[Dossier Médical Partagé](https://www.dmp.fr/)** est largement critiqué. Cette application gère le suivi médical et l'identité des patients, une sorte de carnet de santé virtuel. En adhérant aux principes zero knowledge on obtient exactement le controle que l'on souhaite avoir sur ses données avec la sécurité induite par l'architecture. L'utilisateur a un compte sur lequel se trouve tout son historique médical numérisé, ses rendez-vous, ses ordonnances etc... Grâce à ce compte il partage sélectivement ses informations avec les professionnels de santé, les ordonnances avec les pharmaciens, les informations de facturation avec sa mutuelle et chaque professionel ne voit que les données qui lui sont utiles. Compte tenu de la criticité des données gérées c'est exactement le genre de scénaro pour lequel la ZKA a beaucoup d'interet.

Reviewer 2: ```suggestion
C'est un cas d'application idéal pour l'architecture ZKA, d'autant plus l'applicatif actuel du **[Dossier Médical Partagé](https://www.dmp.fr/)** est largement critiqué. Cette application gère le suivi médical et l'identité des patients, une sorte de carnet de santé virtuel. En adhérant aux principes zero knowledge on obtient exactement le contrôle que l'on souhaite avoir sur ses données avec la sécurité induite par l'architecture. L'utilisateur a un compte sur lequel se trouve tout son historique médical numérisé, ses rendez-vous, ses ordonnances etc... Grâce à ce compte il partage sélectivement ses informations avec les professionnels de santé, les ordonnances avec les pharmaciens, les informations de facturation avec sa mutuelle et chaque professionnel ne voit que les données qui lui sont utiles. Compte tenu de la criticité des données gérées c'est exactement le genre de scénario pour lequel la ZKA a beaucoup d’intérêt.
```

### Autres service ZKA

Au cours de nos recherches pour la rédaction de notre article, nous avons pu identifier d'autres service qui utilise les principes zero-knowledge dans leurs produits. On pourra citer la messagerie chiffrée **[Signal](https://signal.org/blog/private-contact-discovery/)**, le gestionnaire de mots de passge **[NordPass](https://nordpass.com/features/zero-knowledge-architecture/)**, ou encore le gestionnaire de documents en ligne **[CryptPad](https://blog.cryptpad.fr/2017/02/20/Time-to-Encrypt-the-Cloud/)** (le lien vers l'article CryptPad fournit une liste d'autres services zero knowledge).

## Conclusion

### Points forts de la ZKA

Cette architecture généralise l'usage du chiffrement des données et de la cryptographie, cela apporte un  bénéfice double :  un renforcement de la sécurité des données à caractères privé, et l'éducation des utilisateurs à l'usage de la cryptographie.

L'écosystème de bibliothèque de cryptographie permet de développer tout types de clients : client lourds, mobiles ou Web, pour les services zero-knowledge.

> La systématisation du principe de défi cryptographique, implique que les échanges de mot de passe sont limités au fournisseur de preuve, ce qui tend à diminuer le risque de fuite de mot de passe.

Reviewer 2: ```suggestion
La systématisation du principe de défi cryptographique implique que les échanges de mot de passe soient limités au fournisseur de preuve, ce qui tend à diminuer le risque de fuite.
```

De plus, le challenge émis par le serveur étant différent à chaque requête, les tentatives d'usurpation de types « man in the middle » sont aussi rendues plus difficiles à opérer.

> Préservation de la vie privée avec l'approche non-naïve, en ne divulguant que les données nécessaires. Certains service Zero-knowledge proposent même un contrôle dans le temps d'accès aux données de l'utiliateur, en mettant en place des mécanismes de révocation et d'expiration des clefs qui permetent de déchiffrer ses données.

Reviewer 2: ```suggestion
Préservation de la vie privée avec l'approche non-naïve, en ne divulguant que les données nécessaires. Certains service Zero-knowledge proposent même un contrôle dans le temps d'accès aux données de l’utilisateur, en mettant en place des mécanismes de révocation et d'expiration des clefs qui permettent de déchiffrer ses données.
```

### Limites

Compte tenu de l'omniprésence de la cryptographie dans l'architecture, la ZKA est assez complexe, ce qui implique une montée en compétence sur la cryptographie et un coût de mise en place supplèmentaire.

Les mécanismes de cryptographies utiliseés par la ZKA sont plus complexes à opérer que les mécanismes classiques. En effet, la cryptographie moderne implique de savoir gérer une clef privée. Par exemple, pour s'assurer de conserver l'accès à des données, il faut mettre en place un système de récupération de clé, là où classiquement on utiliserait une ré-initialisation de mot de passe.

 Attention toutefois à ne pas généraliser cette difficulté. En effet, la récupération de données n'est pas forcément critique pour tous les scénarios, par exemple un service de messagerie instantanée  pourra se permettre de perde l'historique des conversations (et on pourra négliger la récupération des clefs privées utilisées pour le chiffrement des messages).
 À contrario, il est inconcevable de perdre l'accès à ses données médicales.

> Pour développer un client ZKA le plus sûr possible dans les navigateurs on a besoin de nombreuses mesures de sécurités, CORS, CSP (Content security policy), SRI (verification d'assets), Referrer-Policy et File-API. Afin de complexifier les manipulations lors de l'exécution il est préferable de faire tourner la couche crypto en WebAssembly car le javascript peut être manipulé à la volée .

Reviewer 2: ```suggestion
Pour développer un client ZKA le plus sûr possible dans les navigateurs on a besoin de nombreuses mesures de sécurités, [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS), [CSP (Content security policy)](https://developer.mozilla.org/fr/docs/Web/HTTP/CSP), [SRI (verification d'assets)](https://developer.mozilla.org/fr/docs/Web/Security/Subresource_Integrity), [Referrer-Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy) et [File-API](https://www.w3.org/TR/FileAPI/). Afin de complexifier les manipulations lors de l'exécution il est préferable de faire tourner la couche crypto en WebAssembly car le javascript peut être manipulé à la volée .
```
